compaction/compaction.cc

```diff
@@ -156,17 +172,35 @@ static api::timestamp_type get_max_purgeable_timestamp(const table_state& table_
         if (compacting_set.contains(sst)) {
             continue;
         }
+        api::timestamp_type min_timestamp = sst->get_stats_metadata().min_timestamp;
+        auto ts_stats = sst->get_ext_timestamp_stats();
+        if (!ts_stats.empty()) {
+            auto maybe_set_min_timestamp = [&] (sstables::ext_timestamp_stats_type t) {
+                if (auto it = ts_stats.find(t); it != ts_stats.end()) {
+                    min_timestamp = it->second;
+                    return true;
+                }
+                return false;
+            };
+            if (is_shadowable) {
+                if (!maybe_set_min_timestamp(sstables::ext_timestamp_stats_type::min_live_row_marker_timestamp)) {
+                    maybe_set_min_timestamp(sstables::ext_timestamp_stats_type::min_live_timestamp);
```

Comment:
Nit: IIUC MV updates are always written with a row marker - and IIUC we either write both `min_live_row_marker_timestamp` and `min_live_timestamp` to the sstable or none of them - so I don't see when this fallback could ever kick in.

Does it make sense to include it?

Reply:
This is kind of defensive, as I didn't want the code to be that tightly coupled to a decision made somewhere else and that might change in the future, so the fallback just provides a safety net.

That said, we can decide that the lack of `min_live_row_marker_timestamp` can mean that there are not live row markers in the sstable, and treat it like api::max_timestamp (which is the stored in this series in this case)

Reply:
Shadowable tombstones are currently only used for materialized views, and as far as I know, they were designed specifically for them. I don't think that the behavior of a shadowable tombstone is defined if there is no corresponding row marker. Therefore, it was a bit confusing to me to see code which explicitly considers `min_live_timestamp` in case of shadowable tombstones, that seemed to imply something about their semantics and didn't look like a fallback.

I agree that it's good to be defensive here, so I won't insist on changing the code. Perhaps reorganizing the code slightly would help.

As for safety, I originally thought about removing the fallback to `min_live_timestamp` and let the code fall back to the existing `min_timestamp` which should be more conservative and even safer in some sense.

> That said, we can decide that the lack of `min_live_row_marker_timestamp` can mean that there are not live row markers in the sstable, and treat it like api::max_timestamp (which is the stored in this series in this case)

Personally, I think it is not worth skipping a few of bytes per sstable to store the timestamp and be explicit, not to confuse the "not set" and "api::max_timestamp" cases.

Reply:
Simplified in next version by calling `on_internal_error_noexcept` if the required stat is missing from ts_stats. This will abort in testing, and in release mode just fall back to the legacy min_timestamp.